// cassette/dist/external/soundcloud/Song.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const Song_1 = require("../../core/Song");
class SoundcloudSong extends Song_1.default {
    constructor(service, track, playlistID) {
        super(service);
        this.type = 'soundcloud';
        this.title = track.title;
        this.trackID = track.id;
        this.streamURL = track.stream_url;
        this.playlistID = playlistID;
    }
    stream() {
        return __awaiter(this, void 0, void 0, function* () {
            const svc = this.service;
            const result = yield svc.request.get(this.streamURL, {
                responseType: 'stream',
            });
            return result.data;
        });
    }
    next() {
        return __awaiter(this, void 0, void 0, function* () {
            return null;
        });
    }
}
exports.default = SoundcloudSong;
