// cassette/dist/external/youtube/Song.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const ytdl = require("ytdl-core");
const Song_1 = require("../../core/Song");
class YouTubeSong extends Song_1.default {
    constructor(service, video, playlistID) {
        super(service);
        this.type = 'youtube';
        this.title = video.title;
        this.trackID = video.id;
        this.streamURL = video.url;
        this.playlistID = playlistID;
    }
    stream() {
        return ytdl(this.streamURL, {
            filter: 'audioonly',
            quality: 'lowest',
        });
    }
    next() {
        return __awaiter(this, void 0, void 0, function* () {
            return null;
        });
    }
}
exports.default = YouTubeSong;
